bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeConverter.java

```diff
@@ -317,8 +317,10 @@ public static Type getConvertibleType(Object inputValue, Type targetType, String
                 return getConvertibleFiniteType(inputValue, (BFiniteType) targetType, varName,
                         errors, unresolvedValues, allowNumericConversion);
             case TypeTags.TYPE_REFERENCED_TYPE_TAG:
-                return getConvertibleType(inputValue, ((BTypeReferenceType) targetType).getReferredType(), varName,
-                        unresolvedValues, errors, allowNumericConversion);
+                Type referredType = ((BTypeReferenceType) targetType).getReferredType();
+                Type convertibleType = getConvertibleType(inputValue, referredType, varName, unresolvedValues, errors,
+                        allowNumericConversion);
+                return referredType == convertibleType ? targetType : convertibleType;
```

Comment:
Yes, I think. For each `BTypeReferenceType`, I believe the `referredType` is not `null`. So when the `convertibleType` is `null`, it will return `null`.